AssertionError in test_append_large()

For some reason the test started to fail on master:
```
    def test_append_large(tempfile):
        word = "supercalifragilisticexpialidocious"
        n = 10000
        DT = dt.Frame([word] * n, names=["..."])
        DT.to_csv(tempfile)
        DT.to_csv(tempfile, append=True, method="write")
        assert readfile(tempfile) == "...\n" + (word + "\n") * (2*n)
>       DT.to_csv(tempfile, append=True, method="mmap")
E       AssertionError: Assertion 'src && data_' failed in src/core/writebuf.cc, line 211
```

--- a/src/core/writebuf.cc
+++ b/src/core/writebuf.cc
@@ -208,8 +208,8 @@ void ThreadsafeWritableBuffer::write_at(size_t pos, size_t n, const void* src) {
       "length " << n << ", however the buffer is allocated for " << allocsize_
       << " bytes only";
   }
-  XAssert(src && data_);
   dt::shared_lock<dt::shared_mutex> lock(shmutex_, /* exclusive = */ false);
+  XAssert(src && data_);  // data_ should be accessed under the mutex
   char* target = static_cast<char*>(data_) + pos;
   std::memcpy(target, src, n);
 }

